### PR TITLE
fix: remove pointless synchronization in layout cache

### PR DESCRIPTION
**Pull request by SUPERCILEX**
The cache is a thread-local, so doesn't make much sense to require synchronized initialization.

### CODE REVIEW

File: src/layout/layout.rs
Comment (SUPERCILEX) [reply src/layout/layout.rs:41]: Yeah I had a similar thought but the API seemed to be designed to allow initializing a zero capacity cache for people who don't want one. As long as we don't care about that, I see no reason to not just have a cache always present.

Comment (SUPERCILEX) [reply src/layout/layout.rs:41]: Hmmm, though actually this would still need a RefCell. Also `init_cache` is now breaking because the returned bool doesn't make sense.

Comment (joshka) [reply src/layout/layout.rs]: No-one is using init_cache in the wild (at least on a public GitHub repo), so this is safe to break.

https://github.com/search?q=ratatui+init_cache+AND+%28NOT+is%3Afork%29+AND+%28NOT+org%3Aratatui-org%29&type=code

File: src/layout/layout.rs
Comment (joshka) [inline src/layout/layout.rs:218]: It's worth keeping a note about the default size in the doc comment

File: src/layout/layout.rs
Comment (SUPERCILEX) [reply src/layout/layout.rs]: Lol that works :)

File: src/layout/layout.rs
Comment (joshka) [inline src/layout/layout.rs:214]: This can be simplified to `LAYOUT_CACHE.set()`, or alternatively with_borrow() and then calling [resize()](https://docs.rs/lru/latest/lru/struct.LruCache.html#method.resize) on the cache. Not sure which would be better. I suppose the resize approach means that we leave anything already cached in place. I'm likely bikeshedding at this point though.

I think also since we're breaking this already, do you think it would be worth moving the NonZeroUsize to the method signature here. This would make constructing a valid value the app's problem and not ours.

File: src/layout/layout.rs
Comment (joshka) [inline src/layout/layout.rs:563]: I'd extract (area, self.clone() into a variable named key. You end up with more expressive code and even save a line as the block no longer wraps.

File: src/layout/layout.rs
Comment (joshka) [inline src/layout/layout.rs:1086]: Same here - the borrow can be moved to  `LAYOUT_CACHE.with_borrow()`

File: src/layout/layout.rs
Comment (joshka) [inline src/layout/layout.rs:1096]: Probably don't need to test this twice now - this test was to show that it didn't change on the second call. Now that it does, this seems unnecessary.

File: src/layout/layout.rs
Comment (SUPERCILEX) [reply src/layout/layout.rs]: No no, this is great. I had looked for `reserve` and `set_capacity` which aren't a thing so I thought you couldn't resize the cache.

> I think also since we're breaking this already, do you think it would be worth moving the NonZeroUsize to the method signature here. This would make constructing a valid value the app's problem and not ours.

Yup, seems better for users to find out zero ain't valid at compile time rather than crashing later.

File: src/layout/layout.rs
Comment (joshka) [inline src/layout/layout.rs:40]: This is strange that this didn't get caught by the format lint as too wide. Maybe due to being in a macro?

**Review by joshka — APPROVED** [review]

**Comment by joshka** [comment]
I'll let one of the other maintainers weigh in on the breaking change here.

Generally we keep a good log of these in the BREAKING CHANGES doc, which I'm almost inclined to skip on this one, but if you were interested, this could be as simple here as "Layout::init_cache no longer returns bool and takes a NonZeroUsize instead of usize".

File: src/layout/layout.rs
Comment (SUPERCILEX) [reply src/layout/layout.rs]: Yeah, fmt never seems to work right in macros.

**Comment by SUPERCILEX** [comment]
> Generally we keep a good log of these in the BREAKING CHANGES doc, which I'm almost inclined to skip on this one, but if you were interested, this could be as simple here as "Layout::init_cache no longer returns bool and takes a NonZeroUsize instead of usize".

I copied you verbatim. :grin:

**Pull request merged**
